### PR TITLE
Fixes for VS2015 and CMake 3.0

### PR DESCRIPTION
**Pull request by cdglove**
- Use full paths to include files and pch files.
- Link to output lib for MSVC 2015
- Add missing CMake 3.0.2+ flags.

There might be a better way of doing this (I'm not a cmake guru), but this changes for things working for me.

### CODE REVIEW

File: CMakePCHCompiler.cmake [hunk 1]
```diff
@@ -72,40 +79,54 @@ function(target_precompiled_header) # target [...] header
...
 				# /Yc - create precompiled header
+				# /Fp - exact location for precompiled header
 				# /Fd - specify directory for pdb output
-				set(flags "/Yc /Fd${pdb_dir}\\")
+				set(flags "/Yc${win_header} /Fp${win_pch} /Fd${pdb_dir}")
+
+				set_source_files_properties(
+					${header}
+					PROPERTIES
+					LANGUAGE ${lang}
```
Comment (cdglove) [inline CMakePCHCompiler.cmake]: Had to have the lang as CXX for VS for some reason or the file would not be built.

File: CMakePCHCompiler.cmake [hunk 2]
```diff
@@ -72,40 +79,54 @@ function(target_precompiled_header) # target [...] header
...
+			get_filename_component(win_pch "${target_dir}/${header}.pch" ABSOLUTE)
+			get_filename_component(win_header "${header}" ABSOLUTE)
 			# /Yu - use given include as precompiled header
 			# /Fp - exact location for precompiled header
 			# /FI - force include of precompiled header
-			set(flags "/Yu${win_header} /Fp${win_pch} /FI${win_header}")
+			target_compile_options(
+				"${target}" PUBLIC "/Yu${win_header}" "/Fp${win_pch}" "/FI${win_header}"
+				)
+			target_link_libraries(${target} ${pch_target})
```
Comment (cdglove) [inline CMakePCHCompiler.cmake]: VC now requires you to link to the output of creating a precompiled header.

File: CMakePCHCompiler.cmake [hunk 3]
```diff
@@ -233,12 +256,21 @@ function(__watch_pch_last_hook variable access value)
...
-					"${pch_target}"
-					PROPERTIES
-					"${property}"
-					"${value}"
-					)
+				if(property STREQUAL "CXX_STANDARD")
+					if(NOT MSVC)
+						target_compile_options(
+							"${pch_target}"
+							PUBLIC "-std=gnu++${value}"
```
Comment (cdglove) [inline CMakePCHCompiler.cmake]: This is a hack. I couldn't find a way to figure out how the actual -std=* string is generated.

**Comment by m4c0** [comment]
Doesn't it means we need to change `cmake_minimum_required` to 3.0.x?

**Comment by onqtam** [comment]
@m4c0 cmake 3.0 is already old so it shouldn't be a problem

**Comment by onqtam** [comment]
I cannot stress enough how important sharing precompiled headers between targets is to me... (500 plugins as .dlls)

Is this project even alive? Add a donate button and I will donate...

**Comment by nanoant** [comment]
It is not really dead, but neither fully alive ;) I have new job and right now I am kind of overwhelmed by my new duties, however I will have a look at these PRs this weekend and see what I can do.
Moreover my solution (this project) is kind of workaround. Proper solution would be to somehow make CMake itself to be aware of precompiled headers, rather use the tricks. This was somehow proposed to KitWare, but never made through.

**Pull request merged**